### ISSUE TITLE
Feature to make sure the envoy container always start and work before the application container launch.

### ISSUE THREAD

**Issue by yws-ss** [aws/aws-app-mesh-controller-for-k8s#489]
**Is your feature request related to a problem?**

According to Kubernetes logical,  all containers in a Pod are being started in parallel and there is no way to define that one container must be started after other container (excluded the initContainer). Dig into the [code](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L924), seems kubernetes launch the containers by order:

```
for _, idx := range podContainerChanges.ContainersToStart {
		start("container", metrics.Container, containerStartSpec(&pod.Spec.Containers[idx]))
	}
```
Currently, the envoy container is run as sidecar container with the application container in the pod, and is injected at the last order of Pod spec automatically . Base on  kubernetes logical, it have some chance that application container start before the proxy container activate. Which cause the application container might have outgoing network issue before the envoy pod start to work.

**Describe the solution you'd like**

From Istio [changelog](https://istio.io/latest/news/releases/1.7.x/announcing-1.7/change-notes/), they roll out a feature that allow user to add and option `holdApplicationUntilProxyStarts`, which causes the sidecar injector to inject the sidecar at the start of the pod’s container list and configures it to block the start of all other containers until the proxy is ready. 

Maybe, we could consider to add the similar function in the AWS app mesh.

**Describe alternatives you've considered**

I can totally understand this might be a very race scenario, moreover, this can be avoid by adding retry mechanism in application logical. However, we still want some feature that is simple to use or configure.

**Comment by cgchinmay** [comment]
@yws-ss Will take a look and get back to you with an update

**Comment by fbuinosquy1985** [comment]
I need this , the Vault Agent container injector fails in combination with the Aws Mesh envoy even if i set on annotations [vault.hashicorp.com/agent-init-first](https://www.vaultproject.io/docs/platform/k8s/injector/annotations#vault-hashicorp-com-agent-init-first)

**Comment by cgchinmay** [comment]
Adding the right folks @jeremymill @Y0Username

**Comment by paulyeo21** [comment]
i realize this post and some of these comments are from a while ago, apologies for the wait for those users that have been waiting for this capability. it has been released on v1.6.0 pls try it by enabling the feature flag:

`--set sidecar.waitUntilProxyReady=true`
